### ISSUE TITLE
RFC: "reference" datasets

### ISSUE THREAD

**Issue by martindurant** [blendernc/blendernc#136]
fsspec's [`ReferenceFileSystem`](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.implementations.reference.ReferenceFileSystem) allows a file system like virtual view onto chunks of bytes stored in arbitrary locations elsewhere, e.g., cloud bucket storage. We can present each byte chunk as a particular path in the filesystem conforming to the zarr hierarchy model, such that the original set of chunks, potentially across many files, appears as a single zarr dataset. This brings the following advantages:
- only zarr (plus the relevant codecs) is required to read the data, but the original data could be locked in HDF5, TIFF or grib2 files (and more to come)
- the metadata is extracted once, so future opening of the dataset does not need to scan through the target files to find metadata, and so the process is much faster
- you get a single logical view over potentially thousands of files, but due to zarr's access model, you only load the data you need
- loading can happen chunk-wise and in parallel

The details of how to make such reference files is described (and this is a work in progress) at https://github.com/intake/fsspec-reference-maker , and the latest result for some 80TB/370k files of water flow modelling can be seen at https://nbviewer.jupyter.org/gist/rsignell-usgs/02da7d9257b4b26d84d053be1af2ceeb . Note that this is using xarray to process the data, but only zarr is needed to load it. The specific procedure for making this reference file will also be published in an upcoming blog post, but [here](https://medium.com/pangeo/fake-it-until-you-make-it-reading-goes-netcdf4-data-on-aws-s3-as-zarr-for-rapid-data-access-61e33f8fe685) is an earlier, smaller example.

**Comment by josuemtzmo** [comment]
BlenderNC does support `zarr` files, this seems more relevant for the `xarray` implementation of `zarr` imports, rather than in this repository. Thus, I will close this issue.

**Comment by martindurant** [comment]
> this seems more relevant for the xarray implementation of zarr imports

Not at all! xarray just happens to be the most common tool for reading zarr for my geoscience-oriented colleagues. ReferenceFileSystem is general, indeed allowing reads of other file types, such as the CSV block reader currently in development. 
The purpose of my project is to give zarr implementations access to non-zarr data like TIFF and HDF without having to implement/install specific readers for those file types.

**Comment by josuemtzmo** [comment]
`BlenderNC` uses `xarray` to read data and although a more general reader sounds interesting, there are many limitations to integrate more file formats than those supported by `xarray` in the current version of BlenderNC.

**Comment by josuemtzmo** [comment]
Fair, I get it now. In fact, I appreciate your insistence on this, it looks quite interesting, as it would allow local rendering using cloud storage, which I've been thinking to implement for a while in BlenderNC. I've reopened the issue, and I will look into implementing a node to using `intake` or `ffspec` to open cloud storage catalogs, following the examples provided. 

```
cat = intake.open_catalog('s3://esip-qhub-public/noaa/nwm/nwm_catalog.yml')
ds = cat['nwm-reanalysis'].to_dask()
```

and/or 
```
fs = fsspec.filesystem('s3', anon=True)
urls = ['s3://' + f for f in fs.glob("s3://noaa-goes16/ABI-L2-SSTF/2020/210/*/*.nc")]
dask.compute(*[dask.delayed(gen_json)(u) for u in urls])
```

It's not clear to me the best or cleanest way implement this, so I will appreciate any recommendation, as it will take some time for me to familiarize myself with  `ffspec`.

**Comment by martindurant** [comment]
I don't know the status of your stack, but Intake and Dask might well be overkill in this case. 

I would start with zarr/fsspec and one particular storage backend of interest, e.g., aiohttp for HTTP. For the cloud stores, you need additional deps: s3 (aiobotocore), gcs (google stack, aiohttp), abfs (azure-*, aiohttp).

Note that you don't need anything in fsspec-reference-maker at load time, only for creating the reference data in the first place; that might be done by someone else. The idea is that this metadata is made once, and then used as many times as needed.
